posts/wit-bringing-types-to-near-contracts.md

```diff
@@ -0,0 +1,310 @@
+---
+title: "`wit`: Bringing types to NEAR smart contracts"
+date: "2022-02-03"
+author: "@willemneal"
+---
+
+# NEAR and WebAssembly
+
+Under the hood, NEAR uses WebAssembly (Wasm) to run smart contracts. While this is
+great, it has some pain points. For normal Wasm binaries, functions exported can
+only have float and integer types. So more complex types like strings have to be
+passed via a pointer, `pass_string(len: u32, ptr: u32)`. This makes it hard to
+tell from the function signature that a string is being passed, it could be
+any binary blob.
+
+For NEAR Wasm binaries this problem is made worse because all export functions have the same
+type signature, `foo()`, that is they take no arguments and return nothing. The reason
+for this is because the arguments are serialized in JSON[note], so the function first
+asks the host for a binary blob and then deserializes it.
+
+So if you can download a contract's binary and inspect it, you'll only have the function names
+with no other information.
+
+## Introducing the `wit` format
+
+[WebAssembly Interface Types](https://hacks.mozilla.org/2019/08/webassembly-interface-types/) seeks to solve the task of passing of more complex types at a low level.
+
+Until this is complete the `wit-bindgen` project created a [`.wit`](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md) format that can be used to generate the needed source "`adapters`" to handle passing the complex types.
+
+For example, in JavaScript strings are encoded as `UTF-16`[note], but is `UTF-8` in Rust. So
+passing the binary blob of the string from a Rust Wasm binary is not a simple copy/pasted.
+
+For our use case we need `remote adapters`.  Whereas `wit-bindgen` expects to pass values
+via a normal function call, we are serializing values and then deserializing them.
+
+## `witme` a CLI tool for generating to and from `.wit`
+
+So after that introduction let's walk through a real example. First `witme` is a Rust binary that can be installed with `cargo`:
+
+```bash
+cargo install witme
+```
+
+Assume that you have a Rust smart contract:
+
+```rust
+use near_sdk::{witgen, near_bindgen}
+
+/// A message that contains some text
+#[witgen]
+pub struct Message {
+  /// Inner string value
+  text: String,
+}
+
+//...
+pub struct Contract {
+  message: Message
+}
+
+#[near_bindgen]
+impl Contract {
+  
+  /// A change call to set the message
+  pub fn set_message(&mut self, message: Message) {
+    self.mesage = message;
+  }
+
+  /// A view call to get the current message
+  pub fn get_message(self) -> Message {
+    self.message
+  }
+}
+
+```
+
+The view method `get_message` returns a `Message` struct. `#[witgen]` is a Rust macro that generates a corresponding [`wit` record](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md#item-record-bag-of-named-fields). See the [witgen repo](https://github.com/bnjjj/witgen)
```

Comment:
The last sentence here seems like an incomplete thought. Maybe you can conclude it with something like:

```suggestion
The view method `get_message` returns a `Message` struct. `#[witgen]` is a Rust macro that generates a corresponding [`wit` record](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md#item-record-bag-of-named-fields). See the [witgen repo](https://github.com/bnjjj/witgen) to learn more about generating `.wit` files from existing code.
```